includes/Modules/Sign_In_With_Google.php

```diff
@@ -53,52 +66,208 @@ final class Sign_In_With_Google extends Module implements Module_With_Assets, Mo
 		$request_method = $this->context->input()->filter( INPUT_SERVER, 'REQUEST_METHOD' );
-
 		if ( 'POST' !== $request_method ) {
 			return;
 		}
 
+		$login_url = wp_login_url();
+
+		// Check if the CSRF token is valid, if not redirect to the login page with an error.
 		$csrf_cookie = $this->context->input()->filter( INPUT_COOKIE, 'g_csrf_token' );
 		$csrf_post   = $this->context->input()->filter( INPUT_POST, 'g_csrf_token' );
-
-		if (
-			! $csrf_cookie ||
-			! $csrf_post ||
-			$csrf_cookie !== $csrf_post
-		) {
-			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_g_csrf_token', wp_login_url() ) );
+		if ( ! $csrf_cookie || $csrf_cookie !== $csrf_post ) {
+			wp_safe_redirect( add_query_arg( 'error', self::INVALID_CSRF_TOKEN_ERROR, $login_url ) );
 			exit;
 		}
 
-		$client_id = $this->get_settings()->get()['clientID'];
-		$id_token  = $this->context->input()->filter( INPUT_POST, 'credential' );
-		try {
-			$client  = new Google_Client( array( 'client_id' => $client_id ) );
-			$payload = $client->verifyIdToken( $id_token );
+		$user = null;
 
-			if ( empty( $payload ) ) {
-				wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+		try {
+			$user = $this->find_or_create_user();
+			if ( is_wp_error( $user ) ) {
+				wp_safe_redirect( add_query_arg( 'error', $user->get_error_code(), $login_url ) );
 				exit;
 			}
-
-			// @TODO implement further flow using $payload in #9339.
-
 		} catch ( \Exception $e ) {
-			wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+			wp_safe_redirect( add_query_arg( 'error', self::INVALID_REQUEST_ERROR, $login_url ) );
+			exit;
+		}
+
+		// Redirect to the error page if the user is not found.
+		if ( ! $user instanceof WP_User ) {
+			wp_safe_redirect( add_query_arg( 'error', self::INVALID_REQUEST_ERROR, $login_url ) );
 			exit;
 		}
+
+		// Redirect to the error page if the user is not a member of the current blog in multisite.
+		if ( is_multisite() ) {
+			$blog_id = get_current_blog_id();
+			if ( ! is_user_member_of_blog( $user->ID, $blog_id ) ) {
+				if ( get_option( 'users_can_register' ) ) {
+					add_user_to_blog( $blog_id, $user->ID, $this->get_default_role() );
+				} else {
+					wp_safe_redirect( add_query_arg( 'error', self::INVALID_REQUEST_ERROR, $login_url ) );
+					exit;
+				}
+			}
+		}
+
+		// Set the user to be the current user.
+		wp_set_current_user( $user->ID, $user->user_login );
+
+		// Set the authentication cookies and trigger the wp_login action.
+		wp_set_auth_cookie( $user->ID );
+		/** This filter is documented in wp-login.php */
+		do_action( 'wp_login', $user->user_login, $user );
+
+		// Use the admin dashboard URL as the redirect URL by default.
+		$redirect_to = admin_url();
+
+		// If we have the redirect URL in the cookie, use it as the main redirect_to URL.
+		$cookie_redirect_to = $this->context->input()->filter( INPUT_COOKIE, 'google_auth_redirect_to' );
```

Comment:
Let's make sure to use `googlesitekit` as the common namespace/prefix for all things global like cookie names, query params, etc.